pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/impl/AvgShedder.java

```diff
@@ -0,0 +1,306 @@
+@Slf4j
+public class AvgShedder implements LoadSheddingStrategy, ModularLoadManagerStrategy {
+    // map bundle to broker.
+    private final Map<BundleData, String> bundleBrokerMap = new HashMap<>();
+    // map broker to Scores. scores:0-100
+    private final Map<String, Double> brokerScoreMap = new HashMap<>();
+    // map broker hit count for high threshold/low threshold
+    private final Map<String, Integer> brokerHitCountForHigh = new HashMap<>();
+    private final Map<String, Integer> brokerHitCountForLow = new HashMap<>();
+
+    // result returned by shedding, map broker to bundles.
+    private final Multimap<String, String> selectedBundlesCache = ArrayListMultimap.create();
+
+    private static final double MB = 1024 * 1024;
+    private static final Random random = new Random();
+
+    public AvgShedder() {
+    }
+
+    @Override
+    public Multimap<String, String> findBundlesForUnloading(LoadData loadData, ServiceConfiguration conf) {
+        selectedBundlesCache.clear();
+        final double minThroughputThreshold = conf.getMinUnloadMessageThroughput();
+        final double minMsgThreshold = conf.getMinUnloadMessage();
+        final double maxUnloadPercentage = conf.getMaxUnloadPercentage();
+
+        final Map<String, Long> recentlyUnloadedBundles = loadData.getRecentlyUnloadedBundles();
+        final double lowThreshold = conf.getLoadBalancerAvgShedderLowThreshold();
+        final double highThreshold = conf.getLoadBalancerAvgShedderHighThreshold();
+        final int hitCountHighThreshold = conf.getLoadBalancerAvgShedderHitCountHighThreshold();
+        final int hitCountLowThreshold = conf.getLoadBalancerAvgShedderHitCountLowThreshold();
+        if (log.isDebugEnabled()) {
+            log.debug("highThreshold:{}, lowThreshold:{}, hitCountHighThreshold:{}, hitCountLowThreshold:{}, "
+                            + "minMsgThreshold:{}, minThroughputThreshold:{}",
+                    highThreshold, lowThreshold, hitCountHighThreshold, hitCountLowThreshold,
+                    minMsgThreshold, minThroughputThreshold);
+        }
+
+        List<String> brokers = new ArrayList<>(loadData.getBrokerData().size());
+        brokerScoreMap.clear();
+        // calculate scores of brokers.
+        for (Map.Entry<String, BrokerData> entry : loadData.getBrokerData().entrySet()) {
+            LocalBrokerData localBrokerData = entry.getValue().getLocalData();
+            String broker = entry.getKey();
+            brokers.add(broker);
+            Double score = calculateScores(localBrokerData, conf);
+            brokerScoreMap.put(broker, score);
+            // collect data to analyze the relations between scores and throughput and messageRate.
+            log.info("broker:{}, scores:{}, throughput:{}, messageRate:{}", broker, score,
+                    localBrokerData.getMsgThroughputIn() + localBrokerData.getMsgThroughputOut(),
+                    localBrokerData.getMsgRateIn() + localBrokerData.getMsgRateOut());
```

Comment:
It's better to improve the logging in this PR. You will log N INFO lines and 1 DEBUG line for N brokers. Since these `BrokerData` objects will be put into `brokers` and sorted, you can just log 1 INFO line for the sorted broker list.

Reply:
Good idea, i have push a new commit, PTAL, thanks.